src/txn/txn_global.c

```diff
@@ -0,0 +1,627 @@
+/*-
+ * Copyright (c) 2014-2017 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_txn_publish_id --
+ *	Publish a transaction's ID.
+ *
+ *	The txn_global->id_rwlock must be held by our caller.
+ */
+void
+__wt_txn_publish_id(WT_SESSION_IMPL *session)
+{
+	WT_TXN *prev, *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_ID) &&
+	    !F_ISSET(txn, WT_TXN_PUBLIC_ID));
+
+	for (prev = TAILQ_LAST(&txn_global->idh, __wt_txn_id_qh);
+	    prev != NULL && WT_TXNID_LT(txn->id, prev->id);
+	    prev = TAILQ_PREV(prev, __wt_txn_cts_qh, idq))
+		;
+	if (prev == NULL)
+		TAILQ_INSERT_HEAD(&txn_global->idh, txn, idq);
+	else
+		TAILQ_INSERT_AFTER(&txn_global->idh, prev, txn, idq);
+	F_SET(txn, WT_TXN_PUBLIC_ID);
+}
+
+/*
+ * __wt_txn_clear_id --
+ *	Clear a transaction's published ID.
+ */
+void
+__wt_txn_clear_id(WT_SESSION_IMPL *session)
+{
+	WT_TXN *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	if (!F_ISSET(txn, WT_TXN_PUBLIC_ID))
+		return;
+
+	__wt_writelock(session, &txn_global->id_rwlock);
+	TAILQ_REMOVE(&txn_global->idh, txn, idq);
+	txn->id = WT_TXN_NONE;
+	__wt_writeunlock(session, &txn_global->id_rwlock);
+	F_CLR(txn, WT_TXN_HAS_ID | WT_TXN_PUBLIC_ID);
+}
+
+/*
+ * __wt_txn_publish_metadata_pinned --
+ *	Publish a transaction's metadata pinned transaction ID.
+ */
+void
+__wt_txn_publish_metadata_pinned(WT_SESSION_IMPL *session)
+{
+	WT_TXN *prev, *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	WT_ASSERT(session, txn->metadata_pinned != WT_TXN_NONE &&
+	    !F_ISSET(txn, WT_TXN_PUBLIC_METADATA_PINNED));
+
+	__wt_writelock(session, &txn_global->metadata_pinned_rwlock);
+	for (prev = TAILQ_LAST(&txn_global->metadata_pinnedh, __wt_txn_id_qh);
+	    prev != NULL &&
+	    WT_TXNID_LT(txn->metadata_pinned, prev->metadata_pinned);
+	    prev = TAILQ_PREV(prev, __wt_txn_mp_qh, metadata_pinnedq))
+		;
+	if (prev == NULL)
+		TAILQ_INSERT_HEAD(
+		    &txn_global->metadata_pinnedh, txn, metadata_pinnedq);
+	else
+		TAILQ_INSERT_AFTER(
+		    &txn_global->metadata_pinnedh, prev, txn, metadata_pinnedq);
+	__wt_writeunlock(session, &txn_global->metadata_pinned_rwlock);
+	F_SET(txn, WT_TXN_PUBLIC_METADATA_PINNED);
```

Comment:
I'm surprised setting this happens after the lock is released - could you add a comment explaining why?